SDXL LCM LoRA (#1639)
huggingface/blog: lcm_sdxl.md

---

> title: "Ultra-Fast SDXL with LoRAs borrowed from Latent Consistency Models"

Reviewer: SDXL in 4 steps with Latent Consistency LoRAs

> title: "Ultra-Fast SDXL with LoRAs borrowed from Latent Consistency Models"

Reviewer: ```suggestion
title: "SDXL in 4 steps with Latent Consistency LoRAs"
```

thumbnail: /blog/assets/lcm_sdxl/thumbnail.jpg
authors:
- user: pcuenq
- user: valhalla

> - user: sayakpaul

Reviewer 2: Put me in the end.

> - user: sayakpaul

Reviewer: ```suggestion
- user: sayakpaul
- user: multimodalart
```

- user: SimianLuo
  guest: true
- user: dg845
  guest: true
---

> # Ultra-Fast SDXL with LoRAs borrowed from Latent Consistency Models

Author: ```suggestion
# SDXL in 4 steps with Latent Consistency LoRAs
```

> Latent Consistency Models (LCM) are a way to decrease the number of steps required to generate a Stable Diffusion (or SDXL) image by _distilling_ the original model into another version that requires fewer steps (4 to 8 instead of the original 25 to 50). Distillation is a type of fine-tuning that attempts to replicate the outputs from a source model using a new one. The distilled model may be designed to be smaller (that’s the case, for instance, of DistilBERT or the recently-released DistilWhisper) or, in this case, require fewer steps to run. It’s usually a lengthy and costly process that requires huge amounts of data, patience, and a few GPUs.

Reviewer 2: ```suggestion
Latent Consistency Models (LCM) are a way to decrease the number of steps required to generate an image with Stable Diffusion (or SDXL) by _distilling_ the original model into another version that requires fewer steps (4 to 8 instead of the original 25 to 50). Distillation is a type of training procedure that attempts to replicate the outputs from a source model using a new one. The distilled model may be designed to be smaller (that’s the case, for instance, of DistilBERT or the recently-released DistilWhisper or SSD-1B) or, in this case, require fewer steps to run. It’s usually a lengthy and costly process that requires huge amounts of data, patience, and a few GPUs.
```

Reviewer 2: Included SSD-1B to keep it a bit more contextual.

> Well, that was the status quo before today!

Reviewer 2: Alright, I am sold.

We are delighted to announce a new method that can essentially make Stable Diffusion and SDXL faster, as if they had been distilled using the LCM process!

> 

Reviewer 2: Maybe tease the readers a bit here bu throwing a sentence shedding light on the realized speed.

## Contents

- [Method Overview](#method-overview)
- [Why does this matter](#why-does-this-matter)
- [Fast Inference with SDXL LCM LoRAs](#fast-inference-with-sdxl-lcm-loras)
  - [Quality Comparison](#quality-comparison)
  - [Guidance Scale and Negative Prompts](#guidance-scale-and-negative-prompts)
  - [Quality vs base SDXL](#quality-vs-base-sdxl)
- [Benchmarks](#benchmarks)
- [LCM LoRAs and Models Released Today](#lcm-loras-and-models-released-today)
- [Bonus: Use LCM LoRAs with regular SDXL LoRAs](#bonus-use-lcm-loras-with-regular-sdxl-loras)
- [How to train LCM LoRAs](#how-to-train-lcm-loras)
- [Resources](#resources)
- [Credits](#credits)

## Method Overview

> So, what’s the trick? The central idea is to use a LoRA trained on a LCM model and apply it to a non-LCM model, together with the denoising schedule of LCM models. If you are itching to see how this looks in practice, just jump to the [next section](#fast-inference-with-sdxl-lcm-loras) to play with the inference code. If you want to train your own LoRAs, this is the process you’d use:

Reviewer 2: ```suggestion
So, what’s the trick? The central idea is to use a LoRA trained on an LCM model and apply it to a non-LCM model, together with the noise scheduler of LCM models. If you are itching to see how this looks in practice, just jump to the [next section](#fast-inference-with-sdxl-lcm-loras) to play with the inference code. If you want to train your own LoRAs, this is the process you’d use:
```

1. Select an available LCM model from the Hub. For example, for SDXL we recommend you start with ….

> 2. Train a LoRA with your desired characteristics (style, dataset) on the LCM model. LoRA is a type of performance-efficient fine-tuning, or PEFT, that is much cheaper to accomplish than full model fine-tuning. For additional details on LoRA training, please check ….

Reviewer 2: Note: please check ... to be updated.

3. Use the LoRA with the standard SDXL diffusion model and the LCM scheduler and bingo, you get high-quality inference in just a few steps.

## Why does this matter?

Fast inference of Stable Diffusion and SDXL enables new use-cases and workflows. To name a few:

> - Accessibility: generative tools can be used effectively by more people, even if they don’t have access to the latest hardware.
> - Faster iteration: get more images and multiple variants in a fraction of the time! This is great for artists and researchers; whether for personal or commercial use.
> - Production workloads may be possible on more accelerators, including CPUs.
> - Cheaper image generation services.

Reviewer 2: ```suggestion
- **Accessibility**: generative tools can be used effectively by more people, even if they don’t have access to the latest hardware.
- **Faster iteration**: get more images and multiple variants in a fraction of the time! This is great for artists and researchers; whether for personal or commercial use.
- Production workloads may be possible on different accelerators, including CPUs.
- Cheaper image generation services.
```

> To frame the speed difference we are talking about, generating a single 1024x1024 image on my M1 Mac with SDXL base takes about a minute. Using the LCM LoRA I get great results in just ~6s (4 steps). This is an order of magnitude faster, and not having to wait for results is a game-changer. Using a 4090 I get almost instant response (less than 1s).

Reviewer 2: ```suggestion
To gauge the speed difference we are talking about, generating a single 1024x1024 image on an M1 Mac with SDXL (base) takes about a minute. Using the LCM LoRA, we get great results in just ~6s (4 steps). This is an order of magnitude faster, and not having to wait for results is a game-changer. Using a 4090, we get almost instant response (less than 1s). This unlocks the use of SDXL in applications where real-time events are a requirement. 
```

## Fast Inference with SDXL LCM LoRAs

The version of `diffusers` released today makes it very easy to use LCM LoRAs:

```py
from diffusers import DiffusionPipeline, LCMScheduler
import torch

model_id = "stabilityai/stable-diffusion-xl-base-1.0"
lcm_lora_id = "latent-consistency/lcm-lora-sdxl"

pipe = DiffusionPipeline.from_pretrained(model_id, variant="fp16")

pipe.load_lora_weights(lcm_lora_id)
pipe.scheduler = LCMScheduler.from_config(pipe.scheduler.config)
pipe.to(device="cuda", dtype=torch.float16)

prompt = "close-up photography of old man standing in the rain at night, in a street lit by lamps, leica 35mm summilux"
images = pipe(
    prompt=prompt,
    num_inference_steps=4,
    guidance_scale=1,
).images[0]
```

Note how the code:
- Instantiates a standard diffusion pipeline with the SDXL 1.0 base model.
- Applies the LCM LoRA.
- Changes the scheduler to the LCMScheduler, which is the one used in latent consistency models.
- That’s it!

This would result in the following full-resolution image:

TODO: image

> 

Reviewer 2: ```suggestion

Below, we note down a couple of important factors that users should keep in mind when working with SDXL LCM.

```

## Benchmarks

This section is not meant to be exhaustive, but illustrative of the generation speed we achieve on various computers. Let us stress again how liberating it is to explore image generation so easily.

| Hardware                               | SDXL LoRA LCM (4 steps) | SDXL standard (25 steps) |
|----------------------------------------|-------------------------|--------------------------|
| Mac, M1 Max                            | 6.5s                    | 64s                      |
| 2080 Ti                                | 4.7s                    | 10.2s                    |
| 3090                                   | 1.4s                    | 7s                       |
| 4090                                   | 0.7s                    | 3.4s                     |
| T4 (Google Colab Free Tier)            | 8.4s                    | 26.5s                    |
| A100 (80 GB)                           | 1.2s                    | 3.8s                     |

> | Intel i9-10980XE CPU (1/36 cores used) | 29s                     | 219s                     |

Reviewer 2: Do we want to provide the code too for full disclosure?

Reviewer 3: Feel free to have it as a notebook on the hub as well

## LCM LoRAs and Models Released Today

- [`latent-consistency/lcm-lora-sdxl`](https://huggingface.co/latent-consistency/lcm-lora-sdxl). LCM LoRA for [SDXL 1.0 base](https://huggingface.co/stabilityai/stable-diffusion-xl-base-1.0), as seen in the examples above.
- [`latent-consistency/lcm-lora-sdv1-5`](https://huggingface.co/latent-consistency/lcm-lora-sdv1-5). LCM LoRA for [Stable Diffusion 1.5](https://huggingface.co/runwayml/stable-diffusion-v1-5).
- [`latent-consistency/lcm-lora-ssd-1b`](https://huggingface.co/latent-consistency/lcm-lora-ssd-1b). LCM LoRA for [`segmind/SSD-1B`](https://huggingface.co/segmind/SSD-1B), a distilled SDXL model that's 50% smaller and 60% faster than the original SDXL.

- [`latent-consistency/lcm-sdxl`](https://huggingface.co/latent-consistency/lcm-sdxl). Full fine-tuned consistency model derived from [SDXL 1.0 base](https://huggingface.co/stabilityai/stable-diffusion-xl-base-1.0).
- [`latent-consistency/lcm-ssd-1b`](https://huggingface.co/latent-consistency/lcm-ssd-1b). Full fine-tuned consistency model derived from [`segmind/SSD-1B`](https://huggingface.co/segmind/SSD-1B).

## Bonus: Use LCM LoRAs with regular SDXL LoRAs

> Using the diffusers + PEFT integration, you can combine LCM LoRAs with regular SDXL LoRAs, giving them the superpower to run LCM inference in only 4 steps.

Reviewer 2: ```suggestion
Using the [diffusers + PEFT integration](https://huggingface.co/docs/diffusers/main/en/tutorials/using_peft_for_inference), you can combine LCM LoRAs with regular SDXL LoRAs, giving them the superpower to run LCM inference in only 4 steps.
```

---
title: "SDXL in 4 steps with Latent Consistency LoRAs"
thumbnail: /blog/assets/lcm_sdxl/thumbnail.png
authors:
- user: pcuenq
- user: valhalla
- user: sayakpaul
- user: multimodalart
- user: SimianLuo
  guest: true
- user: dg845
  guest: true
---

# SDXL in 4 steps with Latent Consistency LoRAs

> Latent Consistency Models (LCM) are a way to decrease the number of steps required to generate an image with Stable Diffusion (or SDXL) by _distilling_ the original model into another version that requires fewer steps (4 to 8 instead of the original 25 to 50). Distillation is a type of training procedure that attempts to replicate the outputs from a source model using a new one. The distilled model may be designed to be smaller (that’s the case, for instance, of DistilBERT or the recently-released DistilWhisper or SSD-1B) or, in this case, require fewer steps to run. It’s usually a lengthy and costly process that requires huge amounts of data, patience, and a few GPUs.

Reviewer 3: making a bit shorter for concreteness
```suggestion
Latent Consistency Models (LCM) are a way to decrease the number of steps required to generate an image with Stable Diffusion (or SDXL) by _distilling_ the original model into another version that requires fewer steps (4 to 8 instead of the original 25 to 50). Distillation is a type of training procedure that attempts to replicate the outputs from a source model using a new one. The distilled model may be designed to be smaller (that’s the case of DistilBERT or the recently-released DistilWhisper) or, in this case, require fewer steps to run. It’s usually a lengthy and costly process that requires huge amounts of data, patience, and a few GPUs.
```

> Latent Consistency Models (LCM) are a way to decrease the number of steps required to generate an image with Stable Diffusion (or SDXL) by _distilling_ the original model into another version that requires fewer steps (4 to 8 instead of the original 25 to 50). Distillation is a type of training procedure that attempts to replicate the outputs from a source model using a new one. The distilled model may be designed to be smaller (that’s the case, for instance, of DistilBERT or the recently-released DistilWhisper or SSD-1B) or, in this case, require fewer steps to run. It’s usually a lengthy and costly process that requires huge amounts of data, patience, and a few GPUs.

Reviewer 4: Maybe link the paper here https://arxiv.org/abs/2310.04378

Well, that was the status quo before today!

We are delighted to announce a new method that can essentially make Stable Diffusion and SDXL faster, as if they had been distilled using the LCM process! How does it sound to run _any_ SDXL model in about 1 second instead of 7 on a 3090, or 10x faster on Mac? Read on for details!

## Contents

- [Method Overview](#method-overview)
- [Why does this matter](#why-does-this-matter)
- [Fast Inference with SDXL LCM LoRAs](#fast-inference-with-sdxl-lcm-loras)
  - [Quality Comparison](#quality-comparison)
  - [Guidance Scale and Negative Prompts](#guidance-scale-and-negative-prompts)
  - [Quality vs base SDXL](#quality-vs-base-sdxl)
  - [LCM LoRAs with other Models](#lcm-loras-with-other-models)
  - [Full Diffusers Integration](#full-diffusers-integration)
- [Benchmarks](#benchmarks)
- [LCM LoRAs and Models Released Today](#lcm-loras-and-models-released-today)
- [Bonus: Combine LCM LoRAs with regular SDXL LoRAs](#bonus-combine-lcm-loras-with-regular-sdxl-loras)
- [How to train LCM LoRAs](#how-to-train-lcm-loras)
- [Credits](#credits)

## Method Overview

> So, what’s the trick? The central idea is to use a LoRA trained on an LCM model and apply it to a non-LCM model, together with the noise scheduler of LCM models. If you are itching to see how this looks in practice, just jump to the [next section](#fast-inference-with-sdxl-lcm-loras) to play with the inference code. If you want to train your own LoRAs, this is the process you’d use:

Reviewer 3: Maybe add a quick parenthesis saying what lora is and linking to a resource

> So, what’s the trick? The central idea is to use a LoRA trained on an LCM model and apply it to a non-LCM model, together with the noise scheduler of LCM models. If you are itching to see how this looks in practice, just jump to the [next section](#fast-inference-with-sdxl-lcm-loras) to play with the inference code. If you want to train your own LoRAs, this is the process you’d use:

Reviewer 4: ```suggestion
So, what’s the trick? 
For latent consistency distillation, each model needs to be distilled separately. The core idea with LCM-LoRA is to add LoRA layers to the trainable model and only train those layers. The resulting LoRAs can then be applied to any fine-tuned version of the model without having to distil them separately. If you are itching to see how this looks in practice, just jump to the [next section](#fast-inference-with-sdxl-lcm-loras) to play with the inference code. If you want to train your own LoRAs, this is the process you’d use:
```

1. Select an available teacher model from the Hub. For example, you can use [SDXL (base)](https://huggingface.co/stabilityai/stable-diffusion-xl-base-1.0), or any fine-tuned or dreamboothed version you like.
2. [Train a LCM LoRA](#how-to-train-lcm-models-and-loras) on the model. LoRA is a type of performance-efficient fine-tuning, or PEFT, that is much cheaper to accomplish than full model fine-tuning. For additional details on PEFT, please check [this blog post](https://huggingface.co/blog/peft) or [the diffusers LoRA documentation](https://huggingface.co/docs/diffusers/training/lora).

> 3. Use the LoRA with the standard SDXL diffusion model and the LCM scheduler and bingo, you get high-quality inference in just a few steps.

Reviewer 3: ```suggestion
3. Use the LoRA with the standard SDXL diffusion model and the LCM scheduler; bingo! You get high-quality inference in just a few steps.
```

> 3. Use the LoRA with the standard SDXL diffusion model and the LCM scheduler and bingo, you get high-quality inference in just a few steps.

Reviewer 4: ```suggestion
3. Use the LoRA with any SDXL diffusion model and the LCM scheduler and bingo, you get high-quality inference in just a few steps.
```

</p>

Need ideas to explore some LoRAs? Check out our experimental [LoRA the Explorer (LCM version)](https://huggingface.co/spaces/latent-consistency/lcm-LoraTheExplorer) Space to test amazing creations by the community and get inspired!

## How to Train LCM Models and LoRAs

As part of the `diffusers` release today, we are providing training and fine-tuning scripts developed in collaboration with the LCM team authors. They allow users to:

- Perform full-model distillation of Stable Diffusion or SDXL models on large datasets such as Laion.
- Train LCM LoRAs, which is a much easier process. As we've shown in this post, it also makes it possible to run fast inference with Stable Diffusion, without having to go through distillation training.

For more details, please check the instructions for [SDXL](https://github.com/huggingface/diffusers/examples/consistency_distillation/README_sdxl.md) or [Stable Diffusion](https://github.com/huggingface/diffusers/examples/consistency_distillation/README.md) in the repo.

We hope these scripts inspire the community to try their own fine-tunes. Please, do let us know if you use them for your projects!

## Credits

> The amazing work on Latent Consistency Models was performed by the [LCM Team](https://latent-consistency-models.github.io), please make sure to check out their code, report and paper. This project is a testament to the enabling power of open source AI, the cornerstone that allows researchers, practitioners and tinkerers to explore new ideas and collaborate. We'd also like to thank [`@madebyollin`](https://huggingface.co/madebyollin) for their continued contributions to the community, including the `float16` autoencoder we use in our training scripts.

Reviewer 4: Little shout out to us :) 

```suggestion
The amazing work on Latent Consistency Models was performed by the [LCM Team](https://latent-consistency-models.github.io) and the [diffusers](https://github.com/huggingface/diffusers) team, please make sure to check out their code, report and paper. This project is a testament to the enabling power of open source AI, the cornerstone that allows researchers, practitioners and tinkerers to explore new ideas and collaborate. We'd also like to thank [`@madebyollin`](https://huggingface.co/madebyollin) for their continued contributions to the community, including the `float16` autoencoder we use in our training scripts.
```

Author: @patil-suraj How about something like this?

```
The amazing work on Latent Consistency Models was performed by the [LCM Team](https://latent-consistency-models.github.io), please make sure to check out their code, report and paper. This project is a collaboration between the [diffusers team](https://github.com/huggingface/diffusers) and the LCM team. We believe it's a testament to the enabling power of open source AI, the cornerstone that allows researchers, practitioners and tinkerers to explore new ideas and collaborate. We'd also like to thank [`@madebyollin`](https://huggingface.co/madebyollin) for their continued contributions to the community, including the `float16` autoencoder we use in our training scripts.
```

Reviewer 2: But also Daniel too no? @pcuenca @patil-suraj

Reviewer 4: Oh yes !

Author: He's one of the authors. Let me know if you'd like a explicit mention here :)

Reviewer 4: In that case, no need.
